clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/ActiveActiveStoreIngestionTask.java

```diff
@@ -491,30 +493,62 @@ protected void processMessageAndMaybeProduceToKafka(
       // call in this context much less obtrusive, however, it implies that all views can only work for AA stores
       int valueSchemaId =
           rmdWithValueSchemaID != null ? rmdWithValueSchemaID.getValueSchemaId() : incomingValueSchemaId;
-      this.viewWriters.forEach(
-          (k, v) -> v.processRecord(
-              mergeConflictResult.getNewValue(),
-              oldValueProvider.get(),
-              keyBytes,
-              versionNumber,
-              incomingValueSchemaId,
-              valueSchemaId,
-              mergeConflictResult.getRmdRecord()));
 
-      // This function may modify the original record in KME and it is unsafe to use the payload from KME directly after
-      // this call.
-      producePutOrDeleteToKafka(
-          mergeConflictResult,
-          partitionConsumptionState,
-          keyBytes,
-          consumerRecord,
-          subPartition,
-          kafkaUrl,
-          kafkaClusterId,
-          beforeProcessingRecordTimestampNs,
-          valueManifestContainer.getManifest(),
-          rmdWithValueSchemaID == null ? null : rmdWithValueSchemaID.getRmdManifest());
+      // Get Futures and
+      // Pass callback with on completion which decrements, once all decrement call producePutOrDeleteToKafka
+      if (this.viewWriters.size() > 0) {
+        Long preprocessingTime = System.currentTimeMillis();
+        CompletableFuture
+            .allOf(
+                this.viewWriters.values()
+                    .stream()
+                    .map(
+                        (v) -> v.processRecord(
+                            mergeConflictResult.getNewValue(),
+                            oldValueProvider.get(),
+                            keyBytes,
+                            versionNumber,
+                            incomingValueSchemaId,
+                            valueSchemaId,
+                            mergeConflictResult.getRmdRecord()))
+                    .collect(Collectors.toList())
+                    .toArray(new CompletableFuture[this.viewWriters.size()]))
```

Comment:
Since this function is called for each write, I prefer that we avoid the Java stream operators which are known to generate a lot of garbage. A simple loop will do, e.g.:

```
      CompletableFuture[] viewWriterFutures = new CompletableFuture[this.viewWriters.size()];
      int index = 0;
      for (VeniceViewWriter viewWriter: this.viewWriters.values()) {
        viewWriterFutures[index++] = viewWriter.processRecord(
            mergeConflictResult.getNewValue(),
            oldValueProvider.get(),
            keyBytes,
            versionNumber,
            incomingValueSchemaId,
            valueSchemaId,
            mergeConflictResult.getRmdRecord());
      }
      CompletableFuture.allOf(viewWriterFutures)...
```

It's not only more efficient but also shorter and less indented to do it the old school way... what's not to love 😅 ?

Reply:
I feel so validated by this.  My brain always wants to think in loops and writing streams is always my least favorite thing to code, but I've had an overwhelming sense that it's what cool coders do, and I should do it too.  Will convert to loop.